test/e2e/framework/job/wait.go

```diff
@@ -118,38 +134,52 @@ func WaitForJobSuspend(ctx context.Context, c clientset.Interface, ns, jobName s
 
 // WaitForJobFailed uses c to wait for the Job jobName in namespace ns to fail
 func WaitForJobFailed(c clientset.Interface, ns, jobName string) error {
-	return wait.PollImmediate(framework.Poll, JobTimeout, func() (bool, error) {
-		curr, err := c.BatchV1().Jobs(ns).Get(context.TODO(), jobName, metav1.GetOptions{})
+	return framework.Gomega().Eventually(context.TODO(), framework.HandleRetry(func(ctx context.Context) (*batchv1.Job, error) {
+		job, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return nil, fmt.Errorf("getting the job in in namespace %s: %w", ns, err)
 		}
+		return job, nil
+	})).WithTimeout(JobTimeout).WithPolling(framework.Poll).Should(framework.MakeMatcher(func(j *batchv1.Job) (func() string, error) {
 
-		return isJobFailed(curr), nil
-	})
+		if isJobCompleted(j) {
+			return func() string {
+				return fmt.Sprintf("job %q completed while waiting for its failure", jobName)
+			}, nil
+		}
+
+		if isJobFailed(j) {
+			return nil, nil
+		}
+
+		return func() string {
+			return fmt.Sprintf("expected job %q to fail", jobName)
+		}, nil
+	}))
 }
 
 // WaitForJobCondition waits for the specified Job to have the expected condition with the specific reason.
 // When the nil reason is passed, the "reason" string in the condition is
 // not checked.
 func WaitForJobCondition(ctx context.Context, c clientset.Interface, ns, jobName string, cType batchv1.JobConditionType, reason *string) error {
-	err := wait.PollUntilContextTimeout(ctx, framework.Poll, JobTimeout, false, func(ctx context.Context) (bool, error) {
-		curr, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
+	return framework.Gomega().Eventually(ctx, framework.HandleRetry(func(ctx context.Context) (*batchv1.Job, error) {
+		job, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return nil, fmt.Errorf("getting the job in namespace %s: %w", ns, err)
 		}
-		for _, c := range curr.Status.Conditions {
+		return job, nil
+	})).WithTimeout(JobTimeout).WithPolling(framework.Poll).Should(framework.MakeMatcher(func(j *batchv1.Job) (func() string, error) {
+		for _, c := range j.Status.Conditions {
 			if c.Type == cType && c.Status == v1.ConditionTrue {
 				if reason == nil || *reason == c.Reason {
-					return true, nil
+					return nil, nil
 				}
 			}
 		}
-		return false, nil
-	})
-	if err != nil {
-		return fmt.Errorf("waiting for Job %q to have the condition %q with reason: %v: %w", jobName, cType, reason, err)
-	}
-	return nil
+		return func() string {
+			return fmt.Sprintf("waiting for Job %q to have the condition %q with reason: %v", jobName, cType, reason)
```

Comment:
also include namespace in the messages (for example using klog.KRef). I remember often doing grep by the namespace when trying to understand some failures. (same comment for other messages)